### PR TITLE
HPCC-29319 Ensure by default PIPE programs are not allowed to run

### PR DESCRIPTION
**Pull request by jakesmith**
Default engines use of 'allowedPipePrograms' to none by default. Add configuration option in helm chart.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
Should there be changes in Roxie/hThor too?

Should we default to allowing roxiepipe?

**Comment by jakesmith** [comment]
> Should there be changes in Roxie/hThor too?

I meant to open a separate JIRA for roxie, as Roxie isn't performing any validation on the pipe process at the moment.
I've now opened : https://track.hpccsystems.com/browse/HPCC-29346

>Should we default to allowing roxiepipe?

By default, the stock environment.xml has had allowedPipePrograms="*", in hthor and nothing in Thor.
Meaning in both cases any pipe program was allowed, so roxiepipe will still be let through after this change in bare-metal,
but yes, maybe we should unconditionally allow roxiepipe by default on all systems.

**Comment by richardkchapman** [comment]
Might be worth another JIRA to discuss defaulting to secure (allow none, or just roxiepipe) rather than insecure (allow all)...

**Comment by jakesmith** [comment]
> Might be worth another JIRA to discuss defaulting to secure (allow none, or just roxiepipe) rather than insecure (allow all)...

In Containerized - this PR now means it defaults to secure (allow none).
That's true in bare-metal too, but the stock environment and I suspect all deployed environments have "*" or no settings, which equates to allow all.

I think it makes sense to extend this PR to always "roxiepipe".

**Comment by jakesmith** [comment]
@richardkchapman - extended to add builtin programs (just 'roxiepipe' for now)

**Review by mckellyln — APPROVED** [review]
I think it looks good.
I think it does now mean bare-metal and containerized have different defaults, but those are easily changed either way.

**Comment by jakesmith** [comment]
@richardkchapman @mckellyln - the pipe tests in the regression suite failed in Thor (they use 'sort').
I had tested queries using pipe programs under Thor in a vanilla environment and they worked.
It may be that the environment.xml used by the regression suite is causing the different.
Investigating.

**Comment by jakesmith** [comment]
> @richardkchapman @mckellyln - the pipe tests in the regression suite failed in Thor (they use 'sort'). I had tested queries using pipe programs under Thor in a vanilla environment and they worked. It may be that the environment.xml used by the regression suite is causing the different. Investigating.

issue result by last commit.

**Comment by mckellyln** [comment]
looks ok

**Review by richardkchapman — APPROVED** [review]

File: helm/hpcc/values.schema.json [hunk 1]
```diff
@@ -1461,6 +1461,9 @@
         },
         "preferredDataReadPlanes": {
           "$ref" : "#/definitions/preferredDataReadPlanes"
+        },
```
Comment (ghalliday) [inline helm/hpcc/values.schema.json:1464]: Also roxie?

File: common/thorhelper/hpccconfig.cpp [hunk 2]
```diff
@@ -90,3 +90,64 @@ void getMemorySpecifications(std::unordered_map<std::string, __uint64> &memorySp
...
+}
+
+void getAllowedPipePrograms(StringBuffer &allowedPrograms, bool addBuiltInPrograms)
+{
+    CriticalBlock block(allowedPipeCS);
+    if (addBuiltInPrograms)
+        allowedPrograms.append(allowedPipeProgramsWithBuiltIns);
+    else
+        allowedPrograms.append(allowedPipePrograms);
+}
```
Comment (ghalliday) [inline common/thorhelper/hpccconfig.cpp:153]: trivial: missing /n

**Review by ghalliday — COMMENTED** [review]
@jakesmith one missing entry in the schema I think, and needs a link to update the documentation (including red-book? since will now default to restricted?)

File: helm/hpcc/values.schema.json (hunk 1 above)
Comment (jakesmith) [reply helm/hpcc/values.schema.json:1464]: yes, I'll add.

File: common/thorhelper/hpccconfig.cpp (hunk 2 above)
Comment (jakesmith) [reply common/thorhelper/hpccconfig.cpp:153]: will fix

**Comment by jakesmith** [comment]
> @jakesmith one missing entry in the schema I think, and needs a link to update the documentation (including red-book? since will now default to restricted?)

Have opened and linked documentation issue: https://track.hpccsystems.com/browse/HPCC-29378

**Comment by jakesmith** [comment]
@ghalliday - regression pipe queries were failing in Roxie, because the stock environment didn't mirror the allowedPipePrograms="*" setting that thor and hthor had.

Now amended, see new commit.

File: common/thorhelper/hpccconfig.cpp [hunk 3]
```diff
@@ -90,3 +90,64 @@ void getMemorySpecifications(std::unordered_map<std::string, __uint64> &memorySp
     }
     memorySpecifications["recommendedMaxMemory"] = recommendedMaxMemory;
 }
+
+static unsigned pipeProgramUpdateHookCBId = 0;
+static const char *builtInPrograms = "roxiepipe"; // csv list
```
Comment (ghalliday) [inline common/thorhelper/hpccconfig.cpp:95]: constexpr?  (no need to change.)

File: common/thorhelper/hpccconfig.cpp [hunk 4]
```diff
@@ -90,3 +90,64 @@ void getMemorySpecifications(std::unordered_map<std::string, __uint64> &memorySp
...
+        Owned<IPropertyTree> config = getComponentConfig();
+        // NB: containerized config supports a different format
+        if (isContainerized())
+        {
+            Owned<IPropertyTreeIterator> iter = config->getElements("allowedPipePrograms");
+            ForEach(*iter)
+                configuredList.appendUniq(iter->query().queryProp(nullptr));
+        }
+        else
+            configuredList.appendListUniq(config->queryProp("@allowedPipePrograms"), ",");
```
Comment (ghalliday) [inline common/thorhelper/hpccconfig.cpp:115]: I think this means old environments will default to deny all.  Should it be "*" if allowedPipePrograms is not set?  Needs a red book item if so.

**Review by ghalliday — APPROVED** [review]
@jakesmith please squash.  Please also see question about backward compatibility

**Comment by jakesmith** [comment]
@ghalliday - squashed.

**Pull request merged**